File: lib/models/reservation.rb
class Reservation
  attr_reader :object_item, :start_date, :end_date

  def initialize(object_item, start_date, end_date)
    @object_item = object_item
    @start_date = start_date
    @end_date = end_date
  end

  def self.create(object_item, start_date, end_date)
    return false unless start_date && end_date && start_date <= end_date

    if object_item.available?(start_date, end_date)
      reservation = new(object_item, start_date, end_date)
      object_item.add_reservation(reservation)
      true
    else
      false
    end
  end
end

File: src/main/java/com/prog5/Reservation.java
package com.prog5;

import java.time.LocalDate;

public record Reservation(Item item, LocalDate startDate, LocalDate endDate) {

    public static boolean canReserve(Item item, LocalDate startDate, LocalDate endDate) {
        for (Reservation reservation : item.getReservations()) {
            if (!(endDate.isBefore(reservation.startDate()) || startDate.isAfter(reservation.endDate()))) {
                return false;
            }
        }
        return true;
    }

    public static boolean reserveItem(Item item, LocalDate startDate, LocalDate endDate) {
        if (canReserve(item, startDate, endDate)) {
            Reservation reservation = new Reservation(item, startDate, endDate);
            item.addReservation(reservation);
            return true;
        }
        return false;
    }
}
